File: 백준/Gold/1753. 최단경로/최단경로.py
import heapq
import sys
input = sys.stdin.readline

INF = int(1e9)
n, m = map(int, input().rstrip().split())
start = int(input().rstrip())
gragh = [[] for _ in range(n+1)]
distance = [INF for _ in range(n+1)]


for _ in range(m):
    u, v, w = map(int, input().rstrip().split())
    gragh[u].append((v, w))
    
def dijkstra(start):
    pq = []
    heapq.heappush(pq, (0, start))
    distance[start] = 0
    
    while pq:
        dist, node = heapq.heappop(pq)
        
        if distance[node] < dist:
            continue
            
        for i in gragh[node]:
            cost = dist + i[1]
            if cost < distance[i[0]]:
                distance[i[0]] = cost
                heapq.heappush(pq, (cost, i[0]))
                
dijkstra(start)

for i in range(1, n+1):
    if distance[i] == INF:
        print('INF')
    else:
        print(distance[i])

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.*;
import java.util.*;

class Node implements Comparable<Node>{
	int idx;
	int cost;

	Node(int idx, int cost) {
		this.idx = idx;
		this.cost = cost;
	}

	@Override
	public int compareTo(Node another) {
		return this.cost - another.cost;
	}
}

public class Main {
	static List<List<Node>> graph = new ArrayList<>();
	static int[] distance;
	static int V, E;

    public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		V = Integer.parseInt(st.nextToken());
		E = Integer.parseInt(st.nextToken());

		distance = new int[V + 1];
		Arrays.fill(distance, Integer.MAX_VALUE);

		for (int i = 0; i < V + 1; i++) {
			graph.add(new ArrayList<>());
		}

		int start = Integer.parseInt(br.readLine());

		for (int i = 0; i < E; i++) {
			String[] nodeInfo = br.readLine().split(" ");
			int u = Integer.parseInt(nodeInfo[0]);
			int v = Integer.parseInt(nodeInfo[1]);
			int w = Integer.parseInt(nodeInfo[2]);
			graph.get(u).add(new Node(v, w));
		}

		dijkstra(start);

		for (int i = 1; i < V + 1; i++) {
			if (distance[i] == Integer.MAX_VALUE) {
				System.out.println("INF");
			} else {
				System.out.println(distance[i]);
			}
		}

	}

	static void dijkstra(int start) {
		Queue<Node> pq = new PriorityQueue<>();
		pq.offer(new Node(start, 0));
		distance[start] = 0;

		while (!pq.isEmpty()) {
			Node cur = pq.poll();

			if (distance[cur.idx] < cur.cost) continue;

			for (Node next : graph.get(cur.idx)) {
				if (distance[next.idx] > distance[cur.idx] + next.cost) {
					distance[next.idx] = distance[cur.idx] + next.cost;
					pq.offer(new Node(next.idx, distance[next.idx]));
				}
			}
		}
	}
}